crates/pg_workspace_new/src/workspace/server/change.rs

```diff
@@ -54,235 +63,278 @@ impl Document {
+        let start_incl = prev_index
+            .map(|i| self.positions[i].1.start())
+            .unwrap_or(start);
+        let end_incl = next_index
+            .map(|i| self.positions[i].1.end())
+            .unwrap_or_else(|| end);
 
-            let affected = new_content
-                .as_str()
-                .get(usize::from(start)..usize::from(end))
-                .unwrap();
+        let end_incl = if is_addition {
+            end_incl.add(diff_size)
+        } else {
+            end_incl.sub(diff_size)
+        };
 
-            // add new statements
-            for range in pg_statement_splitter::split(affected).ranges {
-                let doc_range = range + start;
-                match self
-                    .statements
-                    .binary_search_by(|(_, r)| r.start().cmp(&doc_range.start()))
-                {
-                    Ok(_) => {}
-                    Err(pos) => {
-                        let new_id = self.id_generator.next();
-                        self.statements.insert(pos, (new_id, doc_range));
-                        changed.push(StatementChange::Added(Statement {
-                            ref_: StatementRef {
-                                path: self.path.clone(),
-                                id: new_id,
-                            },
-                            text: new_content[doc_range].to_string(),
-                        }));
-                    }
-                }
-            }
+        let end = if is_addition {
+            end.add(diff_size)
         } else {
-            // get full affected range
-            let mut start = change.range.unwrap().start();
-            let mut end = change.range.unwrap().end();
+            end.sub(diff_size)
+        };
 
-            if end > new_content.text_len() {
-                end = new_content.text_len();
-            }
+        Affected {
+            affected_range: TextRange::new(start, end.min(content_size)),
+            affected_indices,
+            prev_index,
+            next_index,
+            full_affected_range: TextRange::new(start_incl, end_incl.min(content_size)),
+        }
+    }
 
-            for (_, (_, r)) in &affected {
-                // adjust the range to the new content
-                let adjusted_start = if r.start() >= change.range.unwrap().end() {
-                    r.start() + change.diff_size()
-                } else {
-                    r.start()
-                };
-                let adjusted_end = if r.end() >= change.range.unwrap().end() {
-                    if change.is_addition() {
-                        r.end() + change.diff_size()
-                    } else {
-                        r.end() - change.diff_size()
-                    }
+    fn move_ranges(&mut self, offset: TextSize, diff_size: TextSize, is_addition: bool) {
+        self.positions
+            .iter_mut()
+            .skip_while(|(_, r)| offset > r.start())
```

Comment:
We check `offset > r.start()` becuase if the offset is *within* a statements range, that's the modified statement, so we don't have to move it over, right?

Reply:
yes! this is the "easiest" part. anything that's after the changed range will just be moved back. this means that the statement itself stays the same (and the `Statement` which is used to store diagnostics etc). just changing a range is therefore not considered a modification.

Reply:
gotcha, thanks!